elasticdl/pkg/ps/server.go

```diff
@@ -78,80 +78,117 @@ func (s *Server) reportModelVersionIfNeeded(modelVersion int32) {
-	if s.Param.Version > in.CurrentModelVersion {
-		for _, v := range s.Param.NonEmbeddingParam {
-			res.Model.Param = append(res.Model.Param, common.SerializeTensor(v))
+	denseParamPB := make(map[string]*tensor_go_proto.TensorProto)
+	if s.Model.Version > in.Version {
+		for name, tensor := range s.Model.DenseParameters {
+			denseParamPB[name] = tensor.SerializeToTensorProto()
 		}
 	}
-	s.Param.InitStatus = true
-	return &res, nil
+	var resp = proto.PullDenseParametersResponse{
+		Initialized:     true,
+		Version:         s.Model.Version,
+		DenseParameters: denseParamPB,
+	}
+	return &resp, nil
 }
 
-// PullEmbeddingVector pulls embedding vector from server
-func (s *Server) PullEmbeddingVector(ctx context.Context, in *pb.PullEmbeddingVectorRequest) (*pb.Tensor, error) {
+// PullEmbeddingVectors pulls sparse parameter from server
+func (s *Server) PullEmbeddingVectors(ctx context.Context, in *proto.PullEmbeddingVectorsRequest) (*tensor_go_proto.TensorProto, error) {
 	if in.Ids == nil {
-		return &pb.Tensor{}, nil
+		return &tensor_go_proto.TensorProto{}, nil
 	}
-	table := s.Param.GetEmbeddingParam(in.Name)
+	table := s.Model.GetEmbeddingTable(in.Name)
 	if table == nil {
-		return &pb.Tensor{}, fmt.Errorf("Request embedding Table %s not found in Param", in.Name)
+		return &tensor_go_proto.TensorProto{}, fmt.Errorf("Request embedding Table %s not found in Param", in.Name)
 	}
 	t := table.GetEmbeddingVectors(in.Ids)
-	return common.SerializeTensor(t), nil
+	return t.SerializeToTensorProto(), nil
+}
+
+// PushGradients push gradients to server
+func (s *Server) PushGradients(ctx context.Context, in *proto.Model) (*proto.PushGradientsResponse, error) {
+	// TODO: only support async now
+	err := s.Opt.ApplyGradients(in, s.Model)
+	s.versionLock.Lock()
+	s.Model.Version += int32(1)
+	s.versionLock.Unlock()
+	s.reportModelVersionIfNeeded(s.Model.Version)
+	var resp = proto.PushGradientsResponse{
+		Accepted: true,
+		Version:  s.Model.Version,
+	}
+	return &resp, err
 }
 
-// PushModel pushes model to server
-func (s *Server) PushModel(ctx context.Context, in *pb.Model) (*empty.Empty, error) {
+// PushModel push Model to server
+func (s *Server) PushModel(ctx context.Context, in *proto.Model) (*empty.Empty, error) {
 	s.lock.Lock()
 	var err error
-	if !s.Param.InitStatus {
-		err = s.Param.InitFromModelPB(in)
+	if !s.Model.Initialized {
+		err = s.Model.InitFromModelPB(in)
 		if err == nil {
-			s.Param.InitStatus = true
+			s.Model.Initialized = true
 		}
 	}
 	s.lock.Unlock()
 	return &empty.Empty{}, err
 }
 
-// PushEmbeddingInfo pushes embedding info to server
-func (s *Server) PushEmbeddingInfo(ctx context.Context, in *pb.Model) (*empty.Empty, error) {
-	s.lock.Lock()
-	err := s.Param.InitFromModelPB(in)
-	s.lock.Unlock()
-	return &empty.Empty{}, err
+/* DEPRECATED
+// PullVariable pulls variable from server
+func (s *Server) PullVariable(ctx context.Context, in *proto.PullVariableRequest) (*proto.PullVariableResponse, error) {
```

Comment:
If it is DEPRECATED, then we could just remove it.

Reply:
fixed